records/generator.go

```diff
@@ -363,78 +363,100 @@ func (rg *RecordGenerator) listenerRecord(listener string, ns string) {
 
 func (rg *RecordGenerator) taskRecords(sj state.State, domain string, spec labels.Func, ipSources []string) {
 	for _, f := range sj.Frameworks {
-		fname := labels.DomainFrag(f.Name, labels.Sep, spec)
 
-		// insert taks records
-		tail := "." + domain + "."
 		for _, task := range f.Tasks {
 			var ok bool
 			task.SlaveIP, ok = rg.SlaveIPs[task.SlaveID]
 
-			// skip not running or not discoverable tasks
-			if !ok || (task.State != "TASK_RUNNING") {
-				continue
+			// only do running and discoverable tasks
+			if ok && (task.State == "TASK_RUNNING") {
+				rg.taskRecord(task, f, domain, spec, ipSources)
 			}
+		}
+	}
+}
 
-			// define context
-			ctx := struct{ taskName, taskID, slaveID, taskIP, slaveIP string }{
-				spec(task.Name),
-				hashString(task.ID),
-				slaveIDTail(task.SlaveID),
-				task.IP(ipSources...),
-				task.SlaveIP,
-			}
+type context struct {
+	taskName,
+	taskID,
+	slaveID,
+	taskIP,
+	slaveIP string
+}
 
-			// use DiscoveryInfo name if defined instead of task name
-			if task.HasDiscoveryInfo() {
-				ctx.taskName = task.DiscoveryInfo.Name
-			}
+func (rg *RecordGenerator) taskRecord(task state.Task, f state.Framework, domain string, spec labels.Func, ipSources []string) {
 
-			// insert canonical A records
-			canonical := ctx.taskName + "-" + ctx.taskID + "-" + ctx.slaveID + "." + fname
-			arec := ctx.taskName + "." + fname
+	// define context
+	ctx := context{
+		spec(task.Name),
+		hashString(task.ID),
+		slaveIDTail(task.SlaveID),
+		task.IP(ipSources...),
+		task.SlaveIP,
+	}
 
-			rg.insertRR(arec+tail, ctx.taskIP, "A")
-			rg.insertRR(canonical+tail, ctx.taskIP, "A")
+	// use DiscoveryInfo name if defined instead of task name
+	if task.HasDiscoveryInfo() {
+		// LEGACY TODO: REMOVE
```

Comment:
if you're going for backwards compat, then you're missing a line here in the LEGACY block, prior to invoking taskContextRecord: `ctx.taskName = task.DiscoveryInfo.Name`